### PR TITLE
IGNITE-13433 Benchmark confirms operation's latency drop decrease on Cellular switch comparing to PME-free switch

### PR DESCRIPTION
**Pull request by anton-vinogradov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 1]
```diff
@@ -38,33 +41,54 @@ def __init__(self, context, config, java_class_name, params="", timeout_sec=60,
...
+    def stop_node(self, node, clean_shutdown=True, timeout_sec=10):
+        assert node == self.nodes[0]
+        self.stop_async(clean_shutdown)
+        self.await_stopped(timeout_sec)
+
+    def __check_status(self, desired, timeout=1):
+        self.await_event("%s\\|IGNITE_APPLICATION_BROKEN" % desired, timeout, from_the_beginning=True)
+
+        try:
+            self.await_event("IGNITE_APPLICATION_BROKEN", 1, from_the_beginning=True)
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:83]: Can we replace 1 with 0?

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 2]
```diff
@@ -38,33 +41,54 @@ def __init__(self, context, config, java_class_name, params="", timeout_sec=60,
...
+            raise Exception("Java application execution failed. %s" % self.extract_result("ERROR"))
+        except TimeoutError:
+            pass
+
+        try:
+            self.await_event(desired, 1, from_the_beginning=True)
+        except Exception:
+            raise Exception("Java application execution falied.") from None
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:91]: falied -> failed

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 3]
```diff
@@ -38,33 +41,54 @@ def __init__(self, context, config, java_class_name, params="", timeout_sec=60,
         self.servicejava_class_name = servicejava_class_name
         self.java_class_name = java_class_name
         self.timeout_sec = timeout_sec
-        self.stop_timeout_sec = 10
+        self.params = params
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:44]: How is it used?

File: modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py [hunk 4]
```diff
@@ -147,6 +146,8 @@ def __init__(self, modules, **kwargs):
         libs.append("log4j")
         libs = list(map(lambda m: self.path.module(m) + "/*", libs))
 
+        libs.append(IgnitePath(DEV_BRANCH).module("ducktests") + "/*")
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py:149]: Why do you add ducktests to NodeSpec?

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                cnt++;
+
+                Transaction tx = ignite.transactions().txStart();
+
+                cache.put(i, i);
+
+                try {
+                    ((TransactionProxyImpl<?, ?>)tx).tx().prepare(true);
+                }
+                catch (IgniteCheckedException e) {
```
Comment (timoninmaxim) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java:79]: `IgniteAwareApplication::start` already catch any exception and mark app as broken

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/TxStreamer.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import com.fasterxml.jackson.databind.JsonNode;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.cache.affinity.Affinity;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.ducktest.utils.IgniteAwareApplication;
+
+/**
+ *
```
Comment (timoninmaxim) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/TxStreamer.java:31]: Could you please provide a short brief of the class

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.cache.affinity.Affinity;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.ducktest.utils.IgniteAwareApplication;
+import org.apache.ignite.internal.processors.cache.transactions.TransactionProxyImpl;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.transactions.Transaction;
+
+/**
+ *
```
Comment (timoninmaxim) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java:34]: Could you please provide a short brief of the class

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 1 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:83]: 0 cause (may cause?) no-op wait and cause no exception if the statement missed.

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 2 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py]: Fixed

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 3 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:44]: It is used to gain start params of the application.
see cellular_affinity_test.py:155 for example.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py (hunk 4 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py:149]: At this test, we have a special node filter (see CellularAffinityBackupFilter.java) and we have to deploy it to all server node's classpath.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java (hunk 5 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java]: Fixed

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/TxStreamer.java (hunk 6 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/TxStreamer.java:31]: Done

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java (hunk 7 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/PreparedTxStreamer.java]: Done

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.cache.affinity.Affinity;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.ducktest.utils.IgniteAwareApplication;
+
+/**
+ * Streams transactions to specified cell.
+ */
+public class CellularTxStreamer extends IgniteAwareApplication {
+    /**
```
Comment (ivandasch) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java:34]: I suppose that this should be one-lined

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+        waitForActivation();
+
+        IgniteCache<Integer, Integer> cache = ignite.getOrCreateCache(cacheName);
+
+        long[] max = new long[20];
+
+        Arrays.fill(max, -1);
+
+        int key = 10_000_000;
```
Comment (ivandasch) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java:53]: Why 10_000_000? You never use this fact in the code.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            if (!record && cnt > warmup) {
+                record = true;
+
+                initTime = System.currentTimeMillis();
+
+                log.info("Warmup finished");
+            }
+
+            if (record) {
+                for (int i = 0; i < max.length; i++) {
```
Comment (ivandasch) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java:93]: Use PriorityQueue (binary heap) here

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java (hunk 8 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java]: Correct. Fixed.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java (hunk 10 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java:91]: Discussed privately, the array is ok.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java (hunk 9 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/cellular_affinity_test/CellularTxStreamer.java:51]: Fixed

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 1 above)
Comment (timoninmaxim) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:83]: I mean there should be return immediately. As I see this line just checks that previous wait finishes due to error.

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 3 above)
Comment (timoninmaxim) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:44]: ok

File: modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py (hunk 4 above)
Comment (timoninmaxim) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py:149]: There is a code duplication - the same in ApplicationSpec. Could you please move this code to function?

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 1 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:83]: I tried to set 0 during development and this caused error-less behaviour where they were expected.
Seems 0 unacceptable here.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py (hunk 4 above)
Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_spec.py:149]: Seems, we have to plan some refactoring here (at specs), can we postpone this to refactoring?

**Pull request merged**